unittests/AST/ASTImporterFixtures.cpp

```diff
@@ -34,6 +36,53 @@ void createVirtualFileIfNeeded(ASTUnit *ToAST, StringRef FileName,
                                    llvm::MemoryBuffer::getMemBuffer(Code));
 }
 
+void checkImportedSourceLocations(const Decl *FromD, const Decl *ToD) {
+  // Print debug information.
+  const bool Print = false;
+
+  SmallString<1024> ToPrinted;
+  SmallString<1024> FromPrinted;
+  llvm::raw_svector_ostream ToStream(ToPrinted);
+  llvm::raw_svector_ostream FromStream(FromPrinted);
+  ToD->dump(ToStream);
+  FromD->dump(FromStream);
+  std::regex MatchSourceLoc(
```

Comment:
Instead of regex, why can't we use the simple accessors to get the locations and check the preferences of those locations? It would be more straight forward to compare the structures and not their string dump. Also, these tests will fail whenever the format of the dump changes.

Reply:
Also, how is the refactor of this function is related to source location changes?

Reply:
I did not want to make a huge visitor class to check the source locations. And here some kind of "parallel visitor" is needed. Or a similar thing like in AST structural equivalence check. Finding every needed SourceLocation inside an AST node object is a problem too (many kinds of structures should be checked: TypeSourceInfo, template argument lists), there is no universal way for this (except the text dump). Compared to that, this check with dump is much smaller code and every relevant source location for every kind of AST nodes is contained in the dump. It has problems (if not everything from the `From` subtree is imported the check is not good) but better than nothing.

Reply:
Okay for now. But I think when we upstream this then we will have to create that visitor class, because that is way more precise.

Reply:
About the implementation of parallel visitor: We can traverse one Decl with `ASTRecursiveVisitor` with the defult order and put all the visited nodes' locations into a queue. Then we can traverse the second Decl too in the same order and populate a second queue. Then we can compare the elements of the queue.